selfdrive/updated.py

```diff
@@ -1,57 +1,267 @@
+# TODO: consider impact of SIGINT/SIGTERM and whether we should catch and dismount/finalize/etc?
+# TODO: explore doing git gc, and/or limited-depth clones
+
+# Long term roadmap:
+# TODO: test suite to compare merged-to-finalized, even though manual compare looks good now
+# TODO: scons prebuild the update, maybe from manager so it can run offroad-only and be interrupted at onroad
+# TODO: download any required NEOS update in the background
+
+
+import os
 import datetime
 import subprocess
+from stat import *
+import shutil
 import time
+from pathlib import Path
+import fcntl
+from cffi import FFI
 
+from common.basedir import BASEDIR
 from common.params import Params
 from selfdrive.swaglog import cloudlog
 
+STAGING_ROOT = "/data/safe_staging"
+OVERLAY_LOCK = "/tmp/safe_staging_overlay.lock"
+UPDATE_LOCK = "/tmp/safe_staging_update.lock"
+
+OVERLAY_UPPER = os.path.join(STAGING_ROOT, "upper")
+OVERLAY_METADATA = os.path.join(STAGING_ROOT, "metadata")
+OVERLAY_MERGED = os.path.join(STAGING_ROOT, "merged")
+FINALIZED = os.path.join(STAGING_ROOT, "finalized")
+
 NICE_LOW_PRIORITY = ["nice", "-n", "19"]
-def main(gctx=None):
-  params = Params()
+SHORT = os.getenv("SHORT") is not None
 
-  while True:
-    # try network
-    ping_failed = subprocess.call(["ping", "-W", "4", "-c", "1", "8.8.8.8"])
-    if ping_failed:
-      time.sleep(60)
-      continue
+# Workaround for the EON/termux build of Python having os.link removed.
+ffi = FFI()
+ffi.cdef("int link(const char *oldpath, const char *newpath);")
+libc = ffi.dlopen(None)
 
-    # download application update
-    try:
-      r = subprocess.check_output(NICE_LOW_PRIORITY + ["git", "fetch"], stderr=subprocess.STDOUT).decode('utf8')
-    except subprocess.CalledProcessError as e:
-      cloudlog.event("git fetch failed",
-        cmd=e.cmd,
-        output=e.output,
-        returncode=e.returncode)
-      time.sleep(60)
-      continue
-    cloudlog.info("git fetch success: %s", r)
+# **** helper functions ****
 
-    # Write update available param
-    try:
-      cur_hash = subprocess.check_output(["git", "rev-parse", "HEAD"]).rstrip()
-      upstream_hash = subprocess.check_output(["git", "rev-parse", "@{u}"]).rstrip()
-      params.put("UpdateAvailable", str(int(cur_hash != upstream_hash)))
-    except:
-      params.put("UpdateAvailable", "0")
+def wait_between_updates():
+  if SHORT:
+    time.sleep(10)
+  else:
+    time.sleep(60 * 10)
+
+def link(src, dest):
+  # Workaround for the EON/termux build of Python having os.link removed.
+  return libc.link(src.encode(), dest.encode())
+
+def run(cmd, cwd=None):
+  return subprocess.check_output(cmd, cwd=cwd, stderr=subprocess.STDOUT, encoding='utf8')
+
+# **** meaty functions ****
+
+def init_ovfs():
+  if os.path.ismount(OVERLAY_MERGED):
+    # We had to have the overlay lock to get this far, so this must be a stale
+    # mount from a terminated updated process.
+    run(["umount", OVERLAY_MERGED])
```

Comment:
Let's add a `cloudlog.error()` here since this should never happen in practice.

Reply:
Agreed.